### DISCUSSION TITLE
Factory breaking change under consideration....

### DISCUSSION THREAD

**Discussion by hmlongco** [hmlongco/Factory#112] [Announcements]
Issue #111 describes new warnings that can occur when Swift strict-concurrency is set to complete in your project. 

On consideration of the various ramifications, I may need to consider a breaking change to Factory to do this properly. The current definition of SharedContainer is...
```swift
public protocol SharedContainer: ManagedContainer {
    static var shared: Self { get }
}
```
But I may need to change that to...
```swift
public protocol SharedContainer: ManagedContainer {
    static var _container: Self { get set }
}

extension SharedContainer {
    /// Defines a thread safe access mechanism to the shared container type.
    public static var shared: Self {
        get {
            defer { globalRecursiveLock.unlock() }
            globalRecursiveLock.lock()
            return _container
        }
        set {
            defer { globalRecursiveLock.unlock() }
            globalRecursiveLock.lock()
            _container = newValue
        }
    }
}
```
With access to the static shared container managed by access functions. From the outside and in most cases it looks the same...
```swift
let service = Container.shared.myService()
```
With the only issue occurring if people define their own container types.
```swift
public final class MyCustomContainer: SharedContainer {
    /// Define the default shared container.
    public static var _container = MyCustomContainer() // NAME CHANGE NEEDED
    /// Define the container's manager.
    public var manager: ContainerManager = ContainerManager()
    /// Public initializer
    public init() {}
}
```
Any thoughts?

**Comment by o-nnerb** [comment]
Interest issue you are discussing. In my view, I don't think this approach is the most correct.

I was reviewing the Factory code to better understand the bug involving strict-concurrency, and there are two points that can serve as a starting point to address this warning.

The first one is that "shared" should be a "let" property.

The second point is that the Factory needs a PR that implements conformance to Sendable. Even though the code is thread-safe due to the use of locks, Swift is not able to know that at compile time without the use of Sendable.

I believe that by making these improvements and using a constant for "shared," we will have a solution to this warning.

What do you think?

**Comment by hmlongco** [comment]
Ummm... just tried that and it eliminated the majority of the warnings. My biggest problem with that approach was in the fact that for testing and whatnot your couldn't "reset" a shared factory by simply assigning a new one to it.

Those doing so, however, could switch those to simply use a reset...
```swift
// Instead of...
Container.shared = Container()
// Do...
Container.shared.reset()
```
Will noodle some more...

**Reply by o-nnerb** [reply]
Absolutely! It's worth noting that 'shared' is a singleton, so updating its value from elsewhere doesn't really make much sense.

By the way, great solution! The 'reset' approach can be quite useful.

**Reply by hmlongco** [reply]
Pull the develop 2.2 branch if you want to check it.

**Reply by o-nnerb** [reply]
I checked the develop branch here and was able to understand it better. Indeed, if this update in the value of "shared" is intended only to clear what was registered during a unit test, it's much better to use the "reset" method for a general cleanup. Much more intuitive, Michel. Nice!

**Reply by hmlongco** [reply]
If you read the [original documentation on Containers](https://github.com/hmlongco/Factory/blob/main/Sources/Factory/Factory.docc/Basics/Containers.md) you'll see you could always call `container.manager.reset()`, as well as the push and pop mechanism.

You could also do a reassignment to release the container. In effect, the ultimate "reset". The new function is just shorthand to the reset function on the manager.

**Reply by o-nnerb** [reply]
Hmmm, I understand what you're trying to say. Well, I don't think I agree with the idea of being able to update the value of "shared". However, since it's included in the documentation and is a feature of the Factory, I think your initial solution will be the best approach.

As I'm a beginner in using the Factory, I ended up confusing the "reset" method of the manager with your suggestion for the Container. Indeed, it would be strange to have a "reset" method to reproduce the same behavior already implemented for the manager.

Would the idea be to implement a reset like this?

```swift
extension Container {

    func reset() {
        Container.shared = Container()
    }
}
```

Anyway, the concurrency problem wouldn't be solved. Because the warning is related to the property being a var and not a let. So we go back to your initial proposal as the best solution to the problem. 🤔

**Reply by o-nnerb** [reply]
And, it's worth mentioning that changing "shared" to "let" would also introduce breaking changes. 🫠

**Reply by hmlongco** [reply]
Both are breaking changes. The first approach affects those who've created custom containers, the second those who may have done the shared reassignment.

That said, I think the `let` approach is best, as that's easiest to find and replace and doesn't let people bypass the mechanism by accessing "container" directly.

# Factory Changelog

## 2.2.0

* Converts "shared" container variables from var to let. Issue - #111
* Add convenience reset() function to container. Issue - #111

  Note that issue #111 could be a breaking change if your code attempts to "replace" a shared container. Do reset instead.

  Container.shared.reset() // Do this...
  Container.shared = Container() // Not this...

**Reply by jeroenvb-wbd** [reply]
Nice one, I also think using a `let` makes more sense as you are already sure it is thread safe, we can now tell the compiler it is thread safe. 

Using the `reset` method is a nice way to achieve what you want to do, and keeps mutation inside its own class (which can be tested to be thread safe) rather than outside. Thanks for your quick reply and experiments 

Will this be a 3.0 release as it is breaking?

**Reply by hmlongco** [reply]
Major.Minor.Bug

Factory 2.0 was a major change/rewrite, Factory 2.2 will be a minor change that "could" break some code, depending on usage.
